### PR TITLE
search experiment: collapse when hits does not contain my bookmarks

### PR DESCRIPTION
**Pull request by yingjieMiao**

### CODE REVIEW

**Comment by yingjieMiao** [comment]
@ymatsuda

**Comment by ymatsuda** [comment]
Do you want to ignore friends' keeps?

**Comment by yingjieMiao** [comment]
Ah, I see. Misread the spec. Will fix.

File: kifi-backend/modules/search/app/com/keepit/search/MainSearcher.scala [hunk 1]
```diff
@@ -299,6 +301,7 @@ class MainSearcher(
           h.scoring = new Scoring(h.score, score / highScore, bookmarkScore(sharingScore(sharingUsers) + 1.0f), recencyScoreVal, usefulPages.mayContain(h.id, 2))
           h.score = h.scoring.score(myBookmarkBoost, sharingBoostInNetwork, recencyBoost, usefulPageBoost)
           hits.insert(h)
+          numMyHitsToReturn += 1
```
Comment (ymatsuda) [inline kifi-backend/modules/search/app/com/keepit/search/MainSearcher.scala]: Note that this doesn't compute the actual number of my keeps in the final result.

**Comment by ymatsuda** [comment]
For an initial search with the default filter,
1) if we have hits from my keeps or friends' keeps, don't add any hit from others.
2) if no hits from my keeps or friends' keeps, add hits from others and collapse the result.

**Comment by yingjieMiao** [comment]
@ymatsuda 
Please review the new version. Thanks.

File: kifi-backend/modules/search/app/com/keepit/search/MainSearcher.scala [hunk 2]
```diff
@@ -349,32 +350,36 @@ class MainSearcher(
...
-
-      othersHits.toRankedIterator.forall{ case (h, rank) =>
-        val score = h.score * dampFunc(rank, dampingHalfDecayOthers) // damping the scores by rank
-        if (score > othersThreshold) {
-          h.bookmarkCount = getPublicBookmarkCount(h.id) // TODO: revisit this later. We probably want the private count.
-          if (h.bookmarkCount > 0) {
-            h.scoring = new Scoring(h.score, score / othersNorm, bookmarkScore(h.bookmarkCount.toFloat), 0.0f, usefulPages.mayContain(h.id, 2))
-            h.score = h.scoring.score(1.0f, sharingBoostOutOfNetwork, recencyBoost, usefulPageBoost)
-            queue.insert(h)
+      if ( !forbidEmptyFriendlyHits || (forbidEmptyFriendlyHits && hits.size == 0)){
```
Comment (ymatsuda) [inline kifi-backend/modules/search/app/com/keepit/search/MainSearcher.scala]: This also needs
|| !filter.isDefault || !initialSearch

Comment (yingjieMiao) [inline kifi-backend/modules/search/app/com/keepit/search/MainSearcher.scala]: yes, you are right.

**Comment by ymatsuda** [comment]
lgtm

**Pull request merged**